topdown/builtins/builtins.go

```diff
@@ -57,31 +66,28 @@ func (c NDBCache) Get(name string, k ast.Value) (ast.Value, bool) {
 
 // Convenience functions for serializing the data structure.
 func (c NDBCache) MarshalJSON() ([]byte, error) {
-	out := make(map[string]json.RawMessage)
-	for bname, obj := range c {
-		j, err := json.Marshal(ast.NewTerm(obj))
-		if err != nil {
-			return nil, err
-		}
-		out[bname] = j
-	}
-	return json.Marshal(out)
+	return json.Marshal(ast.MustJSON(c.AsValue()))
 }
 
 func (c *NDBCache) UnmarshalJSON(data []byte) error {
 	out := map[string]ast.Object{}
-	var incoming map[string]ast.Term
+	var incoming interface{}
 
 	// We deserialize into a map of Terms, and then extract out the Objects.
-	err := json.Unmarshal(data, &incoming)
+	err := util.Unmarshal(data, &incoming)
 	if err != nil {
 		return err
 	}
-	for k, v := range incoming {
-		if obj, ok := v.Value.(ast.Object); ok {
-			out[k] = obj
-		} else {
+	if source, ok := ast.MustInterfaceToValue(incoming).(ast.Object); ok {
```

Comment:
Those functions can indeed panic. I'll rework this logic, since I can't imagine an OPA-generated cache causing a panic, but a malformed cache coming from a user or malicious party definitely could cause some problems.

Reply:
I have rolled in that change. :+1: